File: 2025/src/day01.rs
fn main() {
    let str = std::fs::read_to_string("data/01.txt").expect("Failed to read input file");
    let input = parse(str);
    println!("part1: {}", part1(&input));
    println!("part2: {}", part2(&input));
}

fn parse(str: String) -> Vec<(i32, i32)> {
    str.lines()
        .filter(|line| !line.is_empty())
        .map(|line| {
            let dir = if line.starts_with('L') { -1 } else { 1 };
            let num = line[1..].parse::<i32>().unwrap();
            (dir, num)
        })
        .collect()
}

fn part1(input: &[(i32, i32)]) -> i32 {
    let mut pos = 50;
    let mut ret = 0;
    input.iter().for_each(|(dir, num)| {
        pos = (pos + dir * num + 100) % 100;
        if pos == 0 {
            ret += 1
        }
    });

    ret
}

fn part2(input: &[(i32, i32)]) -> i32 {
    let mut pos = 50;
    let mut ret = 0;
    input.iter().for_each(|(dir, num)| {
        for _ in 0..*num {
            pos += *dir;
            pos %= 100;
            if pos == 0 {
                ret += 1;
            } else {
            }
        }        
    });

    ret
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let input = parse("L68\nL30\nR48\nL5\nR60\nL55\nL1\nL99\nR14\nL82".to_string());
        assert_eq!(part1(&input), 3);
    }

    #[test]
    fn test_part2() {
        let input = parse("L68\nL30\nR48\nL5\nR60\nL55\nL1\nL99\nR14\nL82".to_string());
        assert_eq!(part2(&input), 6);
    }

    #[test]
    fn test_part2_zeroes() {
        let input = parse("L50\nL100".to_string());
        assert_eq!(part2(&input), 2);
       
    }
}

File: 2017-c/src/day01.c
#include "utils.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int part1(char *input) {
  int ret = 0;

  for (int i = 0; i < strlen(input); i++) {
    char next = input[(i + 1) % strlen(input)];

    if (next == input[i]) {
      ret += input[i] - '0';
    }
  }
  return ret;
}

int part2(char *input) {
  int ret = 0;
  int len = strlen(input);

  for (int i = 0; i < len; i++) {
    char next = input[(i + len / 2) % len];

    if (next == input[i]) {
      ret += input[i] - '0';
    }
  }
  return ret;
}

int main() {
  assert(part1("1122") == 3);
  assert(part1("1111") == 4);
  assert(part1("1234") == 0);
  assert(part1("91212129") == 9);

  assert(part2("1212") == 6);
  assert(part2("1221") == 0);
  assert(part2("123425") == 4);
  assert(part2("123123") == 12);
  assert(part2("12131415") == 4);

  char *input = read_file("../data/01.txt");

  printf("part1: %d\n", part1(input));
  printf("part2: %d\n", part2(input));

  free(input);

  return 0;
}
